File: web/src/knowledge/rag-engine.ts
/**
 * RAG Engine - Retrieval Augmented Generation
 * Combines vector search with LLM to provide context-aware responses
 */

import { IndexedDBVectorStore } from './indexeddb-vector-store';
import { WASMLLMEngine } from '../llm/wasm-llm-engine';
import { BrowserEmbeddingGenerator } from './browser-embeddings';

export interface RAGConfig {
  topK?: number;
  similarityThreshold?: number;
  maxContextLength?: number;
  temperature?: number;
  maxTokens?: number;
}

export interface RAGResponse {
  response: string;
  sources: Array<{
    text: string;
    score: number;
    metadata?: any;
  }>;
  contextUsed: string;
}

export class RAGEngine {
  private vectorStore: IndexedDBVectorStore;
  private llmEngine: WASMLLMEngine;
  private embeddingGenerator: BrowserEmbeddingGenerator;
  private config: Required<RAGConfig>;

  constructor(
    vectorStore: IndexedDBVectorStore,
    llmEngine: WASMLLMEngine,
    config: RAGConfig = {}
  ) {
    this.vectorStore = vectorStore;
    this.llmEngine = llmEngine;
    this.embeddingGenerator = new BrowserEmbeddingGenerator(384); // Default dimension
    
    this.config = {
      topK: config.topK ?? 3,
      similarityThreshold: config.similarityThreshold ?? 0.7,
      maxContextLength: config.maxContextLength ?? 2000,
      temperature: config.temperature ?? 0.7,
      maxTokens: config.maxTokens ?? 512
    };
  }

  /**
   * Process a query using RAG
   */
  async processQuery(
    query: string,
    options: { stream?: boolean; filter?: { source?: string; category?: string } } = {}
  ): Promise<RAGResponse | AsyncGenerator<string, void, unknown>> {
    // For MVP: Use placeholder embedding (will be replaced with actual embedding model)
    // TODO: Implement actual embedding generation or use pre-computed query embeddings
    const queryEmbedding = await this.embeddingGenerator.generate(query);

    // Search for relevant documents
    const searchResults = await this.vectorStore.search(
      queryEmbedding,
      this.config.topK,
      options.filter
    );

    // Filter by similarity threshold
    const relevantDocs = searchResults.filter(
      result => result.score >= this.config.similarityThreshold
    );

    if (relevantDocs.length === 0) {
      // No relevant context found, use LLM without context
      return this.generateWithoutContext(query, options.stream || false);
    }

    // Build context from retrieved documents
    const context = this.buildContext(relevantDocs);

    // Generate response with context
    if (options.stream) {
      return this.generateStreaming(query, context);
    } else {
      return this.generate(query, context, relevantDocs);
    }
  }

  /**
   * Build context string from retrieved documents
   */
  private buildContext(docs: Array<{ document: any; score: number }>): string {
    let context = '';
    let totalLength = 0;

    for (const result of docs) {
      const docText = result.document.text;
      const docLength = docText.length;

      if (totalLength + docLength > this.config.maxContextLength) {
        // Truncate if needed
        const remaining = this.config.maxContextLength - totalLength;
        context += docText.substring(0, remaining) + '\n\n';
        break;
      }

      context += docText + '\n\n';
      totalLength += docLength;
    }

    return context.trim();
  }

  /**
   * Generate response with context
   */
  private async generate(
    query: string,
    context: string,
    sources: Array<{ document: any; score: number }>
  ): Promise<RAGResponse> {
    const prompt = this.buildPrompt(query, context);

    const llmResponse = await this.llmEngine.generate(prompt, {
      temperature: this.config.temperature,
      max_tokens: this.config.maxTokens
    });

    return {
      response: llmResponse.text,
      sources: sources.map(s => ({
        text: s.document.text.substring(0, 200) + '...', // Preview
        score: s.score,
        metadata: s.document.metadata
      })),
      contextUsed: context
    };
  }

  /**
   * Generate streaming response with context
   */
  private async *generateStreaming(
    query: string,
    context: string
  ): AsyncGenerator<string, void, unknown> {
    const prompt = this.buildPrompt(query, context);

    for await (const chunk of this.llmEngine.generateStream(prompt, {
      temperature: this.config.temperature,
      max_tokens: this.config.maxTokens
    })) {
      yield chunk;
    }
  }

  /**
   * Generate response without context (fallback)
   */
  private async generateWithoutContext(
    query: string,
    stream: boolean
  ): Promise<RAGResponse | AsyncGenerator<string, void, unknown>> {
    const prompt = `You are dant, a helpful AI assistant. Answer the following question based on your training data:\n\nQuestion: ${query}\n\nAnswer:`;

    if (stream) {
      return this.llmEngine.generateStream(prompt, {
        temperature: this.config.temperature,
        max_tokens: this.config.maxTokens
      });
    } else {
      const response = await this.llmEngine.generate(prompt, {
        temperature: this.config.temperature,
        max_tokens: this.config.maxTokens
      });

      return {
        response: response.text,
        sources: [],
        contextUsed: ''
      };
    }
  }

  /**
   * Build prompt with context
   */
  private buildPrompt(query: string, context: string): string {
    return `You are dant, a helpful AI assistant with access to a knowledge base. Use the following context to answer the question. If the context doesn't contain relevant information, use your general knowledge.

Context from knowledge base:
${context}

Question: ${query}

Answer based on the context above:`;
  }

  /**
   * Update configuration
   */
  updateConfig(config: Partial<RAGConfig>): void {
    this.config = { ...this.config, ...config };
  }
}

File: src/agent/rag_engine.py
"""RAG (Retrieval Augmented Generation) engine."""

from typing import List, Dict, Optional, Any
from ..knowledge.vector_store import VectorStore
from ..knowledge.embeddings import EmbeddingGenerator
from ..config.settings import get_settings


class RAGEngine:
    """Retrieval Augmented Generation engine for accessing knowledge bank."""
    
    def __init__(
        self,
        vector_store: Optional[VectorStore] = None,
        embedding_generator: Optional[EmbeddingGenerator] = None,
        top_k: Optional[int] = None,
        similarity_threshold: Optional[float] = None
    ):
        """
        Initialize RAG engine.
        
        Args:
            vector_store: Vector store instance. If None, creates new one.
            embedding_generator: Embedding generator instance. If None, creates new one.
            top_k: Number of top results to retrieve. If None, uses default from config.
            similarity_threshold: Minimum similarity threshold. If None, uses default from config.
        """
        settings = get_settings()
        self.top_k = top_k or settings.get("rag.top_k", 3)
        self.similarity_threshold = similarity_threshold or settings.get("rag.similarity_threshold", 0.7)
        
        # Initialize components
        self.vector_store = vector_store or VectorStore()
        self.embedding_generator = embedding_generator or EmbeddingGenerator()
    
    def retrieve(
        self,
        query: str,
        top_k: Optional[int] = None,
        filter_metadata: Optional[Dict] = None
    ) -> List[Dict[str, Any]]:
        """
        Retrieve relevant documents from knowledge bank.
        
        Args:
            query: Query text
            top_k: Number of results to return. If None, uses instance default.
            filter_metadata: Optional metadata filter for retrieval
            
        Returns:
            List of retrieved documents with 'text', 'metadata', 'distance' keys
        """
        k = top_k or self.top_k
        
        # Generate query embedding
        query_embedding = self.embedding_generator.encode(query)
        
        # Search vector store
        results = self.vector_store.search(
            query_embeddings=[query_embedding],
            n_results=k,
            where=filter_metadata
        )
        
        # Format results
        retrieved_docs = []
        if results["ids"] and len(results["ids"][0]) > 0:
            for i in range(len(results["ids"][0])):
                doc_id = results["ids"][0][i]
                doc_text = results["documents"][0][i] if results["documents"] else ""
                metadata = results["metadatas"][0][i] if results["metadatas"] else {}
                distance = results["distances"][0][i] if results["distances"] else 1.0
                
                # Convert distance to similarity (cosine distance -> similarity)
                similarity = 1.0 - distance
                
                # Filter by similarity threshold
                if similarity >= self.similarity_threshold:
                    retrieved_docs.append({
                        "id": doc_id,
                        "text": doc_text,
                        "metadata": metadata,
                        "similarity": similarity,
                        "distance": distance
                    })
        
        return retrieved_docs
    
    def format_context(
        self,
        retrieved_docs: List[Dict[str, Any]],
        max_length: int = 2000
    ) -> str:
        """
        Format retrieved documents into context string for LLM.
        
        Args:
            retrieved_docs: List of retrieved documents
            max_length: Maximum length of context string (characters)
            
        Returns:
            Formatted context string
        """
        if not retrieved_docs:
            return ""
        
        context_parts = []
            current_length = 0
        
        for doc in retrieved_docs:
            doc_text = doc["text"]
            doc_length = len(doc_text)
            
            if current_length + doc_length > max_length:
                break
            
            # Add document with metadata if available
            if doc.get("metadata"):
                source = doc["metadata"].get("source", "knowledge base")
                context_parts.append(f"[Source: {source}]\n{doc_text}")
            else:
                context_parts.append(doc_text)
            
            current_length += doc_length + 2  # +2 for newline
        
        return "\n\n".join(context_parts)
    
    def query(
        self,
        query: str,
        top_k: Optional[int] = None,
        max_context_length: int = 2000
    ) -> str:
        """
        Retrieve and format context for a query.
        
        Args:
            query: Query text
            top_k: Number of results to retrieve. If None, uses instance default.
            max_context_length: Maximum length of formatted context
            
        Returns:
            Formatted context string
        """
        retrieved_docs = self.retrieve(query, top_k=top_k)
        context = self.format_context(retrieved_docs, max_length=max_context_length)
        return context
